crates/commitlog/src/commit.rs

```diff
@@ -0,0 +1,233 @@
+    ///
+    /// - Or, the read bytes are all zeroes
+    ///
+    ///   This is to allow preallocation of segments.
+    ///
+    pub fn decode<R: Read>(mut reader: R) -> io::Result<Option<Self>> {
+        let mut hdr = [0; Self::LEN];
+        if let Err(e) = reader.read_exact(&mut hdr) {
+            if e.kind() == io::ErrorKind::UnexpectedEof {
+                return Ok(None);
+            }
+
+            return Err(e);
+        }
+        match &mut hdr.as_slice() {
+            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] => Ok(None),
+            buf => {
+                let min_tx_offset = buf.get_u64().map_err(decode_error)?;
+                let n = buf.get_u16().map_err(decode_error)?;
+                let len = buf.get_u32().map_err(decode_error)?;
+
+                Ok(Some(Self { min_tx_offset, n, len }))
+            }
+        }
+    }
+}
+
+/// Entry type of a [`crate::Commitlog`].
+#[derive(Debug, Default, PartialEq)]
+pub struct Commit {
+    /// The offset of the first record in this commit.
+    ///
+    /// The offset starts from zero and is counted from the beginning of the
+    /// entire log.
+    pub min_tx_offset: u64,
+    /// The number of records in the commit.
+    pub n: u16,
+    /// A buffer of all records in the commit in serialized form.
+    ///
+    /// Readers must bring their own [`crate::Decoder`] to interpret this buffer.
+    /// `n` indicates how many records the buffer contains.
+    pub records: Vec<u8>,
+}
+
+impl Commit {
+    pub const FRAMING_LEN: usize = Header::LEN + /* crc32 */ 4;
+    pub const CHECKSUM_ALGORITHM: u8 = CHECKSUM_ALGORITHM_CRC32C;
+
+    /// The largest transaction offset in this commit.
+    pub fn max_tx_offset(&self) -> u64 {
+        self.min_tx_offset + self.n as u64
+    }
+
+    /// Length in bytes of this commit when written to the log via [`Self::write`].
+    pub fn encoded_len(&self) -> usize {
+        Self::FRAMING_LEN + self.records.len()
+    }
+
+    /// Serialize and write `self` to `out`.
+    pub fn write<W: Write>(&self, out: W) -> io::Result<()> {
+        let mut out = Crc32cWriter::new(out);
+
+        let min_tx_offset = self.min_tx_offset.to_le_bytes();
+        let n = self.n.to_le_bytes();
+        let len = (self.records.len() as u32).to_le_bytes();
+
+        out.write_all(&min_tx_offset)?;
+        out.write_all(&n)?;
+        out.write_all(&len)?;
+        out.write_all(&self.records)?;
+
+        let crc = out.crc32c();
+        let mut out = out.into_inner();
+        out.write_all(&crc.to_le_bytes())?;
+
+        Ok(())
+    }
+
+    /// Attempt to read one [`Commit`] from the given [`Read`]er.
+    ///
+    /// Returns `None` if the reader is already at EOF.
+    ///
+    /// Verifies the checksum of the commit. If it doesn't match, an error of
+    /// kind [`io::ErrorKind::InvalidData`] with an inner error downcastable to
+    /// [`ChecksumMismatch`] is returned.
+    pub fn decode<R: Read>(reader: R) -> io::Result<Option<Self>> {
```

Comment:
We could try to support reading into a user-supplied buffer / zero-copy reads with mmap in the future, but that should be left for later.

Reply:
True. Gotta get some benchmarks in place...